gl3w_gen.py

```diff
@@ -253,31 +253,114 @@ def download(url, dst):
 #else
 #include <dlfcn.h>
 
-static void *libgl;
-static GL3WglProc (*glx_get_proc_address)(const GLubyte *);
+static void *libgl;  // OpenGL library
+static void *libglx;  // GLX library
+static void *libegl;  // EGL library
+static GL3WGetProcAddressProc gl_get_proc_address;
 
-static int open_libgl(void)
+static void close_libgl(void)
+{
+	if (libgl) {
+		dlclose(libgl);
+		libgl = NULL;
+	}
+	if (libegl) {
+		dlclose(libegl);
+		libegl = NULL;
+	}
+	if (libglx) {
+		dlclose(libglx);
+		libglx = NULL;
+	}
+}
+
+static int is_library_loaded(const char *name, void **lib)
+{
+	*lib = dlopen(name, RTLD_LAZY | RTLD_LOCAL | RTLD_NOLOAD);
+	return *lib != NULL;
+}
+
+static int open_libs(void)
 {
+	// On Linux we have two APIs to get process addresses: EGL and GLX.
+	// EGL is supported under both X11 and Wayland, whereas GLX is X11-specific.
+
+	libgl = NULL;
+	libegl = NULL;
+	libglx = NULL;
+
+	// First check what's already loaded, the windowing library might have
+	// already loaded either EGL or GLX and we want to use the same one.
+
+	if (is_library_loaded("libEGL.so.1", &libegl) ||
+			is_library_loaded("libGLX.so.0", &libglx)) {
```

Comment:
What's the purpose of this `libGLX` check? Is it just an alternative to using GLX via `libGL`? (And if so, why is it being checked with `libEGL` instead of below with `libGL`?)

Reply:
Yes, it's an alternative to `libGL`.
On modern Linux systems there are two ways to load GLX:

 * The new GLVND way: load GLX from libGLX and OpenGL from libOpenGL.
 * The legacy way: load both GLX and OpenGL from libGL.  On modern systems this just routes calls to libOpenGL or libGLX (or `libGLdispatch`, but that's a GLVND implementation detail).

See [the GLVND README](https://gitlab.freedesktop.org/glvnd/libglvnd/-/blob/master/README.md) and [OpenGL ABI probosal](https://github.com/aritger/linux-opengl-abi-proposal/blob/master/linux-opengl-abi-proposal.txt) that it links to for more information.

The reason the check is here is because if we load `libGLX` we also have to load `libOpenGL`, same as with `libEGL`.  If `libEGL` is found the condition will short-circuit and it won't check for `libGLX`, otherwise it will check for `libGLX`.  Inside the conditional body exactly one of `libegl` or `libglx` will be non-null.  Either way we'll want to attempt to load `libOpenGL` to complete the EGL/GLX + OpenGL pairing.

Reply:
Apparently I forgot to respond, thanks for the explanation!